test/extended/util/cli.go

```diff
@@ -183,41 +187,58 @@ func (c *CLI) Verbose() *CLI {
 	return c
 }
 
-// Client provides an OpenShift client for the current user. If the user is not
-// set, then it provides client for the cluster admin user
-func (c *CLI) Client() *client.Client {
+func (c *CLI) AppsClient() appsclientset.Interface {
 	_, clientConfig, err := configapi.GetInternalKubeClient(c.configPath, nil)
-	osClient, err := client.New(clientConfig)
+	client, err := appsclientset.NewForConfig(clientConfig)
 	if err != nil {
 		FatalErr(err)
 	}
-	return osClient
+	return client
 }
 
-// Clientset provides the new OpenShift clientset for the current user. If the user is not
-// set, then it provides client for the cluster admin user
-func (c *CLI) Clientset() *oclientset.Clientset {
```

Comment:
@deads2k @mfojtik why did we remove this? Having all the methods directly on CLI object seems wrong to me. What if we need them in other packages? Are we to write them all over again?
This was nicely abstracted in https://github.com/openshift/origin/blob/4cf8f8e04eccde0163fc935748e62a31a08cf87c/pkg/client/clientset/clientset/clientset.go#L1-L61

Reply:
also now I have to write AppsClient().Apps() instead of Clientset().Apps() which made more sense to me than writing Apps twice...

Reply:
Manual clientsets are no longer allowed.  Only generated clientsets are allowed going forward.



> Are we to write them all over again?

I don't think we're ever going to allow it again.

Reply:
@deads2k this was a wrapper for **generated clientset**

Reply:
something like client-go until that is a thing

Reply:
> @deads2k this was a wrapper for generated clientset

That's not a thing we want or need for the same reasons we rejected making a combined openshift and kube client.  There's no need for a dependency magnet tying those pieces together.  It's just a place that takes maintenance.  If we end up generating a unified clientset for other reasons (which would be undesireable, but perhaps required for golang vendoring reasons), then it will come back, but I'd still discourage it.

Reply:
@deads2k I think this PR proves exactly such need -  you have created N function on CLI and duplicated probably all of them with admin prefix creating 2*N functions. It would have been just 2 functions if we would used the pattern I have started (copying client-go's design).

I understand that internally in our code to keep dependencies low separate clients are desirable. But this "grouped" design is the exact pattern that's in client-go so I fail to see why that's bad given that it is in addition to the separated clientsets.

It is also consistent with how we use upstream client - you get kubernetes clientsets and you choose API group by say '.CoreV1()'. Not having our APIs in a similar manner looks inconsistent.

Reply:
> I understand that internally in our code to keep dependencies low separate clients are desirable. But this "grouped" design is the exact pattern that's in client-go so I fail to see why that's bad given that it is in addition to the separated clientsets.

At the simplest level: because different APIs grow at different rates.  oauth.openshift.io isn't related to image.openshift.io.  They can and probably will, version themselves at different rates.  When you bundle them together, you're forcing a lock-step migration that is unnecessary and bundling two unrelated things.  Manually tying it together doesn't make your life easier.

Reply:
not even considering you need those duplicated for versioned clients

Reply:
> not even considering you need those duplicated for versioned clients

Internal and external should always be separate.

Reply:
> At the simplest level: because different APIs grow at different rates. 

I fail to see this to be an issue. You can add methods for different API groups independently into the wrapper:
```go
Core()
CoreV1()
Route()
RouteV1()
RouteV2alpha1()
```

Still this is the design in client-go. @deads2k are you saying that's doing it wrong?

As a user I want to have a client that can talk to OpenShift and I am fine having a separate Kubernetes client. But why do I have to create 10 separate clients for every openshift object? I don't have to do it in Kubernetes and they have multiple API groups. So why in OpenShift?

Reply:
> Still this is the design in client-go. @deads2k are you saying that's doing it wrong?

Yes.  Emphatically, yes!  They have coupled together APIs that deal with jobs and how to build an extensible API server together.  Doing so, they've bound together multiple unrelated APIs.

Reply:
I think I have to agree to disagree on this.

If I have a project integrating with OpenShift I need to interact with multiple API groups. A very simplified example would be a function to create a namespace, route, service and DC. That's at least 3 different OpenShift API groups and if you need to move those 3 (usually more) separate clientsets around almost every user ends up wrapping it in a struct - just like client-go does it for e.g. CoreV1(), AppsV1()...
https://github.com/kubernetes/client-go/blob/master/kubernetes/clientset.go#L50

I was strongly hoping that our client would provide such unified interface, not forcing every user to write this on his own :/